File: lib/platform/platform.py
"""Parent class for all Platforms.

The Platform class implements push and query, but leaves enum
for the individual platforms to implement.
"""

from argparse import Namespace
from colorama import init
from jinja2 import Environment, FileSystemLoader
import json
import logging
from neo4j import GraphDatabase, Query
from neo4j.exceptions import AuthError, ClientError, TransientError, ServiceUnavailable
import os
import re
from lib.helpers.helpers import Modes
import sys
from tqdm import tqdm
from termcolor import colored
import yaml

logger = logging.getLogger('konstellation')

# init colorama
init()

class Platform(object):
    """Platform base class."""
    def __init__(self, config: Namespace, **kwargs) -> None:
        self.config = config

        # assing all kwargs as class members
        for key, value in kwargs.items():
            setattr(self, key, value)

        if self.config.mode in [Modes.push.name, Modes.query.name]:
            # Load and validate mappings
            platform_config_file = os.path.join('resources',
                                                self.name, 'config.yml')
            if not os.path.exists(platform_config_file):
                raise FileNotFoundError(f'{self.name} ' \
                                        f'{platform_config_file} ' \
                                        'config file not found.')

            logger.debug('Loading %s', platform_config_file)
            with open(platform_config_file, 'r', encoding='ascii') as fin:
                platform_config = yaml.safe_load(fin.read())
            self.mappings = platform_config['mappings']
            self.order = platform_config.get('order', {})
            self.relationships = platform_config['relationships']
            self.queries = platform_config['queries']

            if not 'default' in self.mappings:
                raise KeyError(f'{platform_config_file} is missing' \
                               ' a "default" template mapping entry')


            # Set up Jinja2
            self.templates = Environment(
                loader=FileSystemLoader(
                os.path.join('resources', self.name)))

            # Create neo4j driver if we're going to perform a push
            self.neo4j = GraphDatabase.driver(self.config.neo4juri,
                                              auth=(self.config.neo4juser,
                                                    self.config.neo4jpass))
            try:
                self.neo4j.verify_connectivity()
            except (ServiceUnavailable, ConnectionRefusedError):
                logger.error('Neo4j unavailable at %s', self.config.neo4juri)
                sys.exit()
            except AuthError as e:
                logger.error('Neo4j AuthError: %s', e)
                sys.exit()

        if self.config.mode == Modes.query.name:
            self._make_dir_if_not_exist(self.config.results)

    @property
    def name(self) -> str:
        return self.__class__.__name__

    def enum(self) -> None:
        raise NotImplementedError(f'Enum for {self.name} has not' \
                                  f'been implemented yet.')

    def push(self) -> None:
        try:
            results = os.listdir(self.config.enum)
        except FileNotFoundError:
            logger.error('%s not found. Run `%s enum` first,' \
                         ' or specify an alternative directory with `--enum`',
                         self.config.enum, self.name)
            sys.exit()

        if not self.config.relationships:
            logger.warning('Pushing %s data from %s into %s.',
                           self.name,
                           os.path.abspath(self.config.enum),
                           self.config.neo4juri)
            ordered = self._order_items(results)
            for filename in tqdm(ordered):
                # render abs path for result file so neo4j can locate it
                path = os.path.join(os.path.abspath(self.config.enum), filename)
                template_name = self._get_template_name(filename)
                template_file_name = self._get_template_file_name(template_name)

                logger.info('Importing %s using template %s.',
                            filename, template_name)
                if self.mappings[template_name].get('labelField'):
                    label = self.mappings[template_name].get('labelField')
                else:
                    label = f"'{self.mappings[template_name].get('label')}'"

                data = {
                    'path': path,
                    'jsonPath': self.mappings[template_name].get('jsonPath', '$'),
                    'label': label,
                    'labelField': self.mappings[template_name].get('labelField', None),
                    'nameField': self.mappings[template_name].get('nameField', '$') 
                }

                logger.debug('data: %s', data)
                cypher = self._render_template(template_file_name, **data)
                logger.debug(cypher)
                try:
                    self._neo4j_query(cypher)
                except ClientError as e:
                    logger.error('Error importing %s', filename)
                    logger.debug(e)
                except TransientError as e:
                    logger.error(e)

        self._apply_relationship_mappings()

    def query(self) -> None:
        for q in self.queries:
            if q.get('template', None):
                query = self._render_template(q.get('template'))
            else:
                query = q.get('query')

            if self.config.name not in [None ,q.get('name')]:
                continue

            logger.warning('Executing query: %s', q.get('name'))
            res = self._neo4j_query(query)

            filename = self._make_filename(q.get('name'), 'json')
            path = os.path.join(self.config.results, filename)

            data = None
            try:
                data = json.dumps(res, indent=2)
            except TypeError:
                # only fields were returned
                data = str(res)

            if self.config.print:
                for i in res:
                    tqdm.write(str(i))

            if len(res) > 0:
                logger.critical(colored(f'Found {len(res)} results', 'green'))
                logger.warning('Writing %s "%s" results to %s.',
                               len(res), q.get('name'), path)
                with open(os.path.abspath(path), 'w', encoding='ascii') as fout:
                    fout.write(data)
            else:
                logger.warning('Found 0 results')

    def _make_filename(self, value: str, extension: str) -> str:
        """Convert query description to a reasonable file name."""
        value = value.lower()
        value = re.sub(r'[^\w\s-]', '', value)
        value = re.sub(r'[-\s]+', '-', value)
        return f'{value}.{extension}'

    def _neo4j_query(self, cypher: str) -> list:
        """Execute the cypher query."""
        with self.neo4j.session() as session:
            logger.debug(cypher)
            q = Query('') # workaround for LiteralString error
            q.text = cypher

            res = []
            try:
                res = session.run(q)
            except ClientError as e:
                logger.error('Error executing cypher query: %s', e)
                return []

            # Results belong to the session,
            # so create a list of values to return
            return [ dict(i) for i in res]

    def _get_template_name(self, filename: str) -> str:
        if self.mappings.get(filename, None):
            template = filename
        else:
            template = 'default'

        logger.debug('filename: %s, template: %s', filename, template)
        return template

    def _get_template_file_name(self, template_name: str) -> str:
        template = self.mappings.get(template_name).get('template', 'generic.cypher')

        logger.debug('template_name: %s, template: %s', template_name, template)
        return template

    def _make_dir_if_not_exist(self, dir_: str) -> None:
        if not os.path.exists(dir_):
            os.makedirs(dir_)

    def _render_template(self, template_name: str, **kwargs) -> str:
        template = self.templates.get_template(template_name)
        rendered = template.render(**kwargs)
        return rendered

    def _order_items(self, items: list) -> list:

        if self.order.get('last', None):
            items = [item for item in items if item not in self.order['last']]
            items = items + self.order['last']
        return items

    def _apply_relationship_mappings(self) -> None:
        logger.info('Applying relationships, this may take a while.')
        query = ''
        relationships = []

        if self.config.relationship_name:
            for r in self.relationships:
                if self.config.relationship_name == r['name']:
                    relationships.append(r)
                    break
            if len(relationships) != 1:
                logger.error('Relationship %s not found.', self.config.relationship_name)
                sys.exit()
        else:
            relationships = self.relationships

        for r in tqdm(relationships):
            data = {}
            if r.get('results_file'):
                path = os.path.join(os.path.abspath(self.config.enum), r['results_file'])
                data = {'path': path}

            query = self._render_template(r['template'], **data)
            logger.warning('Applying %s - %s', r['name'], r.get('description', ''))
            self._neo4j_query(query)

File: pkg/platform/platform.go
package konstellation

import (
	"context"
	"embed"
	"encoding/json"
	"fmt"
	"io"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"strings"
	"sync"

	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
	graphdb "github.com/praetorian-inc/konstellation/pkg/neo4j"
	utils "github.com/praetorian-inc/konstellation/pkg/utils"
	"github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
	"gopkg.in/yaml.v3"
)

type Platform interface {
	Enum()
	Push()
	Query()
}

type BasePlatform struct {
	Name    string
	Config  PlatformConfig
	graphdb graphdb.GraphDB
	cmd     *cobra.Command
}

func (p *BasePlatform) SetCmd(cmd *cobra.Command) {
	p.cmd = cmd
}

func (p *BasePlatform) Push(ctx context.Context) {
	logrus.Info(p.Name + " push")
	//enumOutput, _ := cmd.Flags().GetString("enum")
	relationships, _ := p.cmd.Flags().GetBool("relationships")
	//relationshipName, _ := cmd.Flags().GetString("relationship-name")

	if !relationships {
		fmt.Printf("default: %v", p.Config.Mappings["default"])
		/*if !p.Config.Mappings["default"] {
			configPath := filepath.Join(path, p.Name, "config.yml")
			logrus.Panicf("%v is missing a \"default\" template mapping entry")
			os.Exit(1)
		}*/
		p.pushNodes(ctx)
	}
}

func (p *BasePlatform) pushNodes(ctx context.Context) error {
	runtime.GOMAXPROCS(4)
	logrus.SetLevel(logrus.TraceLevel)
	logrus.SetLevel(logrus.DebugLevel)
	logrus.Debug("Pushing nodes")

	files, err := p.getEnumFiles()
	if err != nil {
		return err
	}

	for _, filePath := range files {

		data, err := os.ReadFile(filePath)
		if err != nil {
			logrus.Errorf("Failed to read file: %v", err)
			continue
		}

		file := filepath.Base(filePath)
		// TODO implement template usage
		//template, _ := p.getMappingValue(file.Name(), "template")
		jsonPath, _ := p.getMappingValue(file, "JsonPath")
		labelField, _ := p.getMappingValue(file, "LabelField")
		label, _ := p.getMappingValue(file, "Label")
		nameField, _ := p.getMappingValue(file, "NameField")

		itemsSlice, err := getItems(data, jsonPath)
		if err != nil {
			logrus.Errorf("Failed to get items: %v", err)
			continue
		}
		logrus.Infof("Parsed JSON file: %v", file)

		var itemLabel string
		var nodeChannel = make(chan graphdb.Node, 1)
		//var errorChannel = make(error, 1)
		var resultsChannel = make(chan any, 1)
		wg := sync.WaitGroup{}

		for i := 0; i < 10; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				for node := range nodeChannel {
					rc, err := p.insertNode(ctx, node)
					if err != nil {
						logrus.Errorf("Failed to insert node: %v", err)
						continue
					}
					resultsChannel <- rc
				}
			}()
		}

		// Iterate over items
		go func() {
			for _, item := range itemsSlice {
				//TODO make the channel size configurable

				//TODO we've got j and item in this loop, which is confusing
				j, err := json.Marshal(item)
				// convert item to map/number of items

				item := item.(map[string]interface{})

				if err != nil {
					logrus.Errorf("Failed to marshal item: %v", err)
					continue
				}

				// get the label
				if label == "" {
					l, err := utils.FilterJson(j, labelField, true)
					if err != nil {
						logrus.Errorf("Failed to get labelField: %s", err)
						continue
					}

					itemLabel = string(l)
				} else {
					itemLabel = label
				}
				logrus.Tracef("Item label: %s", itemLabel)

				name, err := utils.FilterJson(j, nameField, true)
				if err != nil {
					logrus.Errorf("Failed to get nameField: %s", err)
					continue
				}
				logrus.Tracef("Item name: %s", name)
				item["name"] = string(name)

				node := graphdb.Item2Node(item, itemLabel)
				logrus.Trace(node)

				nodeChannel <- node

			}

			close(nodeChannel)
			wg.Wait()
			close(resultsChannel)

		}()

		// consume the results
		for range resultsChannel {
		}
	}

	return nil
}

// getFilesEnum returns a list of file paths for JSON files in the "enum/k8s-enum" directory.
// It uses the provided Platform's cmd field to get the directory path.
// If there is an error while reading the directory or finding JSON files, it returns an error.
func (p *BasePlatform) getEnumFiles() ([]string, error) {

	enumPath, _ := utils.GetDirectoryPath(p.cmd, "enum", "k8s-enum")
	files, err := os.ReadDir(enumPath)
	if err != nil {
		return nil, err
	}

	var fileNames []string
	for _, file := range files {
		if filepath.Ext(file.Name()) == ".json" {
			filePath := filepath.Join(enumPath, file.Name())
			fileNames = append(fileNames, filePath)
		}
	}

	return fileNames, nil
}

// getItems parses the given JSON data using the provided jsonPath and returns a slice of items.
// It filters the JSON data based on the jsonPath, unmarshals the filtered data into an interface{},
// and asserts it to a slice of interfaces. It returns the items slice and any error encountered during the process.
func getItems(data []byte, jsonPath string) (itemsSlice []interface{}, err error) {

	// Parse JSON using jsonPath
	filtered, err := utils.FilterJson(data, jsonPath)
	if err != nil {
		logrus.Errorf("Failed to filter JSON: %v", err)
		return nil, err
	}

	//fmt.Println("jsonPath: ", filtered)
	var items interface{}
	err = json.Unmarshal(filtered, &items)
	if err != nil {
		logrus.Errorf("Failed to parse JSON: %v", err)
		return nil, err
	}

	itemsSlice, ok := items.([]interface{})

	if !ok {
		log.Fatal("Could not assert items to slice")
	}

	logrus.Debug("Number of items: ", len(itemsSlice))

	return itemsSlice, nil
}

func (p *BasePlatform) getMappingValue(fileName string, key string) (string, error) {

	mapping, ok := p.Config.Mappings[fileName]
	if !ok {
		mapping, ok = p.Config.Mappings["default"]
		if !ok {
			logrus.Panic("No default mapping found")
			return "", fmt.Errorf("No default mapping found")
		}
	}

	val := reflect.ValueOf(mapping)
	fieldVal := val.FieldByName(key)
	if !fieldVal.IsValid() {
		return "", fmt.Errorf("Field %v does not exist", key)
	}

	logrus.Tracef("file: %v, key: %v, value: %v", fileName, key, fieldVal)

	return fmt.Sprintf("%v", fieldVal), nil
}

func (p *BasePlatform) insertNode(ctx context.Context, node graphdb.Node) (chan any, error) {
	anyChannel := make(chan any, 1)

	err := p.graphdb.Insert(&node, anyChannel)

	if err != nil {
		logrus.Errorf("Failed to insert node: %v", err)
		return nil, err
	}

	return anyChannel, nil
}

func (p *BasePlatform) applyRelationships() {

	// TODO call generic apply relationships

}

func (p *BasePlatform) Query(ctx context.Context) {
	logrus.Info(p.Name + " query")

	resultChannel := make(chan any, 10)
	var wg sync.WaitGroup

	for _, query := range p.Config.Queries {
		logrus.Info("Running query: " + query.Name)
		logrus.Debug("Query: " + query.Query)

		//TODO write data to file
		// res, err := session...
		err := p.graphdb.Query(query.Query, resultChannel)
		if err != nil {
			logrus.Errorf("Failed to run query: %v", err)
			continue
		}

		wg.Add(1)
		go p.processQueryResults(query, resultChannel, &wg)

		if err != nil {
			return
		}
	}
	close(resultChannel)
	//defer session.Close(ctx)
}

func (p *BasePlatform) loadConfig(fileName string, configPath embed.FS) {
	logrus.Debugf("Using config path %v", fileName)
	config, _ := GetFileContents(fileName, configPath)

	// unmarshal the yaml to the Config var
	err := yaml.Unmarshal(config, &p.Config)
	if err != nil {
		logrus.Fatalf("Failed to unmarshal %v: %v", configPath, err)
	}

	logrus.Debugf("config: %v", p.Config)
}

func NewPlatform(name string, configPath embed.FS) *BasePlatform {
	p := &BasePlatform{
		Name: name,
	}

	return p
}

func (p *BasePlatform) processQueryResults(query Query, resultChannel <-chan any, wg *sync.WaitGroup) {
	defer wg.Done()

	for result := range resultChannel {
		// Cast record to appropriate type
		record := result.(neo4j.Record)

		node := record.Values[0]
		logrus.Trace(node)

		// Convert record to JSON
		jsonData, err := json.Marshal(node)
		if err != nil {
			logrus.Error("Failed to convert record to JSON:", err)
			continue
		}

		// Derive file name from query.Name
		fileName := strings.ReplaceAll(strings.ToLower(query.Name), " ", "_") + ".json"

		// Get results directory from p.cmd
		resultsDir, _ := p.cmd.Flags().GetString("results")
		// Create the directory if it doesn't exist
		err = os.MkdirAll(resultsDir, os.ModePerm)
		if err != nil {
			logrus.Error("Failed to create results directory:", err)
			continue
		}

		// Create the file
		filePath := filepath.Join(resultsDir, fileName)
		file, err := os.Create(filePath)
		if err != nil {
			logrus.Error("Failed to create file:", err)
			continue
		}
		defer file.Close()

		// Write JSON data to the file
		_, err = file.Write(jsonData)
		if err != nil {
			logrus.Error("Failed to write JSON data to file:", err)
			continue
		}
	}
}

func ListFiles(configPath embed.FS) ([]string, error) {
	var files []string

	err := fs.WalkDir(configPath, ".", func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if !d.IsDir() {
			files = append(files, filepath.Join(path))
		}
		return nil
	})

	if err != nil {
		return nil, err
	}

	return files, nil
}

func GetFileContents(fileName string, configPath embed.FS) ([]byte, error) {
	file, err := configPath.Open(fileName)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	contents, err := io.ReadAll(file)
	if err != nil {
		return nil, err
	}

	return contents, nil
}
